Define Action, Event, Properties and ActionRequest resources. Closes #34, #35 and #36

benfrancis opened the pull request:
I've made an attempt to define separate Action, Event and Properties resources as discussed in #82.

I've realised that's there's also technically a separate ActionRequest resource for an individual action request (vs. the Action resource which is a queue of requests for a single action, the Event resource is an event log for a particular event).

This leaves us with separate Properties/Property, Actions/Action/ActionRequest and Events/Event resources and defines where the URLs for each of these resources can be defined.

There are few quirks here (like the separate ActionRequest resource), so feedback is welcome.

You can see a live version at https://benfrancis.github.io/wot/

mrstegeman commented on index.html:
```diff
@@ -381,6 +408,68 @@ <h3><code>Actions</code> resource</h3>
+  {
+    "reboot": {
+      "href": "/things/lamp/actions/reboot/124e4568-f89b-22d3-a356-427656",
+      "timeRequested": "2017-01-24T13:13:33+00:00",
```
This one should also have `timeCompleted`

mrstegeman commented on index.html:
```diff
@@ -381,6 +408,68 @@ <h3><code>Actions</code> resource</h3>
+      <p>The URL of an Action resource can be defined by the <code>href</code> member of an <a href="#action-object">Action object</a> in a Thing Description.</p>
+      <p><strong>Action Request</strong></p>
+      <pre class="example" title="Request an action">
+POST https://mythingserver.com/things/lamp/actions/fade
```
I don't understand the point of this API. It's identical to doing a `POST /things/lamp/actions`.

benfrancis replied:
You're right that it's the same. But in the Thing Description you can provide a href for an action and I felt we needed to define what that URL is used for. A GET on an Action Resource is different to a GET on the top level Actions Resource because it's filtered to a single action, but a POST has the same effect, you're correct. Do you think that's a problem?

mrstegeman commented:
I think if we don't want the POST to be redundant, it should be specified that:
1. Only actions for that `name` are accepted, and/or
2. The extra `name: {}` wrapper need not be included when doing `POST /actions/name`

benfrancis commented:
I've addressed the review comments.

> Only actions for that name are accepted, and/or

I went for this option to try to keep some consistency between HTTP requests and with the WebSocket API which always specifies an action type in an action request.

If another action type is requested, the server should respond with `400 Bad Request`.

mrstegeman commented:
@benfrancis That sounds good. I still had the one comment about the missing `timeCompleted`, but otherwise I'm fine with this.

benfrancis commented:
Yep I fixed that one too (once I'd figured out which branch I was meant to be pushing to!)
